lib/adal-aad-node/src/engine/AzureCodeGrant.ts

```diff
@@ -0,0 +1,104 @@
+import { AzureAuth } from "./AzureAuth";
+import { AzureAuthType, ProviderSettings, SecureStorageProvider, CachingProvider, Logger, MessageDisplayer, ErrorLookup, StringLookup, AADResource, LoginResponse, Tenant, Deferred, AuthorizationCodePostData, OAuthTokenResponse } from "../models";
+import * as crypto from 'crypto';
+import * as qs from 'qs';
+import { AzureAuthError } from "../errors/AzureAuthError";
+
+export class AzureCodeGrant extends AzureAuth {
+    constructor(
+        protected readonly providerSettings: ProviderSettings,
+        protected readonly secureStorage: SecureStorageProvider,
+        protected readonly cachingProvider: CachingProvider,
+        protected readonly logger: Logger,
+        protected readonly messageDisplayer: MessageDisplayer,
+        protected readonly errorLookup: ErrorLookup,
+        protected readonly userInteraction: UserInteraction,
+        protected readonly stringLookup: StringLookup,
+        protected readonly authRequest: AuthRequest,
+    ) {
+        super(providerSettings, secureStorage, cachingProvider, logger, messageDisplayer, errorLookup, userInteraction, stringLookup, AzureAuthType.AuthCodeGrant);
+    }
+
+    protected async login(tenant: Tenant, resource: AADResource): Promise<LoginResponse> {
+        let authCompleteDeferred: Deferred<void>;
+        let authCompletePromise = new Promise<void>((resolve, reject) => authCompleteDeferred = { resolve, reject });
+
+        const { codeVerifier, codeChallenge } = this.createCryptoValues();
+        const state = await this.authRequest.getState();
+
+        const loginQuery = {
+            response_type: 'code',
+            response_mode: 'query',
+            client_id: this.clientId,
+            redirect_uri: this.providerSettings.redirectUri,
+            state,
+            prompt: 'select_account',
+            code_challenge_method: 'S256',
+            code_challenge: codeChallenge,
+            resource: resource.id
+        };
+
+        const signInUrl = `${this.loginEndpointUrl}${tenant}/oauth2/authorize?${qs.stringify(loginQuery)}`;
+        await this.userInteraction.openUrl(signInUrl);
+        const authCode = await this.authRequest.getAuthorizationCode(state);
+
+
+        const response = await this.getTokenWithAuthorizationCode(tenant, resource, {
+            authCode,
+            redirectUri: this.providerSettings.redirectUri,
+            codeVerifier
+        });
+        if (!response) {
+            throw new AzureAuthError(10, this.errorLookup.getSimpleError(10));
```

Comment:
where does this 10 come from?

Reply:
I see now. Adding a comment stating error code 10 is for auth error will help here